src/FSharp Examples/SequenceToSequence.fs

```diff
@@ -0,0 +1,273 @@
+
+    member _.forward(t, mask) =
+        let src = pos_encoder.forward(encoder.forward(t) * sqrEmSz)
+        let enc = transformer_encoder.forward(src, mask)
+        decoder.forward(enc)
+
+    member _.GenerateSquareSubsequentMask(size:int64) =
+        use mask = Float32Tensor.ones([|size;size|]).eq(Float32Tensor.from(1.0f)).triu().transpose(0L,1L)
+        use maskIsZero = mask.eq(Float32Tensor.from(0.0f))
+        use maskIsOne = mask.eq(Float32Tensor.from(1.0f))
+        mask.to_type(ScalarType.Float32)
+            .masked_fill(maskIsZero, Single.NegativeInfinity.ToScalar())
+            .masked_fill(maskIsOne, 0.0f.ToScalar()).``to``(device)
+
+let process_input (iter:string seq) (tokenizer:string->string seq) (vocab:TorchText.Vocab.Vocab) =
+    [|
+        for item in iter do
+            let itemData = [| for token in tokenizer(item) do (int64 vocab.[token]) |]
+            let t = Int64Tensor.from(itemData)
+            if t.NumberOfElements > 0L then
+                t
+    |].cat(0L)
+    
+let batchify (data:TorchTensor) batchSize (device:Device) =
+    let nbatch = data.shape.[0] / batchSize
+    let d2 = data.narrow(0L, 0L, nbatch * batchSize).view(batchSize, -1L).t()
+    d2.contiguous().``to``(device)
+
+let get_batch (source:TorchTensor) (index:int64) =
+
+    let len = min bptt (source.shape.[0]-1L-index)
+    let data = source.[TorchTensorIndex.Slice(index, index + len)]
+    let target = source.[TorchTensorIndex.Slice(index + 1L, index + 1L + len)].reshape(-1L)
+    data,target
+
+let train epoch (model:TransformerModel) (optimizer:Optimizer) (trainData:TorchTensor) ntokens =
+
+    model.Train()
+
+    let mutable total_loss = 0.0f
+    let mutable src_mask = model.GenerateSquareSubsequentMask(bptt)
+
+    let mutable batch = 0
+
+    let tdlen = trainData.shape.[0]
+
+    let mutable i = 0L
+
+    while i < tdlen - 2L  do
+
+        begin
+            let data,targets = get_batch trainData i
+            use data = data
+            use targets = targets
+
+            if data.shape.[0] <> bptt then
+                src_mask.Dispose()
+                src_mask <- model.GenerateSquareSubsequentMask(data.shape.[0])
+
+            optimizer.zero_grad()
+
+            use output = model.forward(data, src_mask)
+            use loss = criterion (output.view(-1L, ntokens)) targets
+            loss.backward()
+            model.parameters().clip_grad_norm(0.5) |> ignore
+            optimizer.step()
+
+            total_loss <- total_loss + loss.cpu().DataItem<float32>()
+        end 
+
+        GC.Collect()
+
+        if (batch % logInterval = 0) && (batch > 0) then
+            let cur_loss = (total_loss / (float32 logInterval)).ToString("0.00")
+            printfn $"epoch: {epoch} | batch: {batch} / {tdlen/bptt} | loss: {cur_loss}"
+            total_loss <- 0.0f
+
+        batch <- batch + 1
+        i <- i + bptt
+
+
+let evaluate (model:TransformerModel) (evalData:TorchTensor) ntokens =
```

Comment:
The mutables + `while` loop in here might be easily translatable into a recursive inner function. That'd be more idiomatic F#, but might not seem as friendly to Python users. What do you think?

Reply:
It would definitely not be Pythonic -- don't know whether that's what we should go for, though.